7_zebra/zebra_pulp.py

```diff
@@ -0,0 +1,126 @@
+     'color': ['red', 'green', 'ivory', 'yellow', 'blue'],
+     'pet': ['dog', 'cat', 'turtle', 'hamster', 'rabbit'],
+     'beverage': ['coffee', 'tea', 'milk', 'orange juice', 'water'],
+     'hobby': ['chess', 'volleyball', 'sudoku', 'karaoke', 'fishing']}
+
+# keys for decision variables x
+x_keys = [(i, j, k) for i in I for j in J for k in K[j]]
+# endregion
+
+# region Define the model
+mdl = pulp.LpProblem('zebra', sense=pulp.LpMaximize)
+
+# add variables
+x = pulp.LpVariable.dicts(indexs=x_keys, cat=pulp.LpBinary, name='x')
+
+# add constraints
+# exactly one option for each house-attribute pair
+for i in I:
+    for j in J:
+        mdl.addConstraint(pulp.lpSum(x[i, j, k] for k in K[j]) == 1, name=f'single_option_{i}_{j}')
+
+# exactly one house for each attribute-option pair
+for j in J:
+    for k in K[j]:
+        mdl.addConstraint(pulp.lpSum(x[i, j, k] for i in I) == 1, name=f'single_house_{j}_{k}')
+
+# R2
+mdl.addConstraint(pulp.lpSum(i * x[i, 'flag', 'Brazil'] for i in I) == pulp.lpSum(i * x[i, 'color', 'red']
+                                                                                     for i in I), name='R2')
+
+# R3
+mdl.addConstraint(pulp.lpSum(i * x[i, 'flag', 'Venezuela'] for i in I) == pulp.lpSum(i * x[i, 'pet', 'dog']
+                                                                                        for i in I), name='R3')
+
+# R4
+mdl.addConstraint(pulp.lpSum(i * x[i, 'color', 'green'] for i in I) == pulp.lpSum(i * x[i, 'beverage', 'coffee']
+                                                                                  for i in I), name='R4')
+# R5
+mdl.addConstraint(pulp.lpSum(i * x[i, 'flag', 'India'] for i in I) == pulp.lpSum(i * x[i, 'beverage', 'tea']
+                                                                                    for i in I), name='R5')
+
+# R6
+for i in I[1:]:
+    mdl.addConstraint(x[i, 'color', 'green'] == x[i - 1, 'color', 'ivory'], name=f'R6_{i}')
+# mdl.addConstraint(x[1, 'color', 'green'] == 0, name=f'R6_{1}')
+
+# R7
+mdl.addConstraint(pulp.lpSum(i * x[i, 'pet', 'cat'] for i in I) == pulp.lpSum(i * x[i, 'hobby', 'chess']
+                                                                              for i in I), name='R7')
+
+# R8
+mdl.addConstraint(pulp.lpSum(i * x[i, 'color', 'yellow'] for i in I) == pulp.lpSum(i * x[i, 'hobby', 'volleyball']
+                                                                                   for i in I), name='R8')
+# R9
+mdl.addConstraint(x[3, 'beverage', 'milk'] == 1, name='R9')
+
+# R10
+mdl.addConstraint(x[1, 'flag', 'Mexico'] == 1, name='R10')
+
+# R11
+for i in I[1:4]:
+    mdl.addConstraint(x[i, 'hobby', 'sudoku'] <= x[i - 1, 'pet', 'turtle'] + x[i + 1, 'pet', 'turtle'],
+                      name=f'R11_{i}')
+
+mdl.addConstraint(x[1, 'hobby', 'sudoku'] <= x[2, 'pet', 'turtle'], name='R11_1')
+
+mdl.addConstraint(x[5, 'hobby', 'sudoku'] <= x[4, 'pet', 'turtle'], name='R11_5')
```

Comment:
In order to avoid the exceptions, there is also the possibility of using these aggregated constraints. However we already know the desegregated version is tighter. This suggestion make sense for all cases where the features must be in the the immediate neighbour.

```suggestion
mdl.addConstraint(
    sum(i * x[i, 'hobby', 'sudoku'] for i in I) - sum(i * x[i, 'pet', 'Turtle'] for i in I) <= 2)
mdl.addConstraint(
       sum(i * x[i, 'hobby', 'sudoku'] for i in I) - sum(i * x[i, 'pet', 'Turtle'] for i in I) >= -2)
```

Reply:
We try to use those in the first version of the implementation,  but there is also an exception.  The aggregated sum can not be 0.

And we fix the number too. Instead of 2 and -2., the correct would be 1 and -1.